text/0088-fix-memory-limitiations-in-session-replays-access-pattern.md

```diff
@@ -0,0 +1,216 @@
+- To eliminate the need for grouping queries we would need to supply the `FINAL` keyword.
+- Our column types are not ideal for this use case.
+
+**Questions**
+
+### 6. Stateful Streaming using Apache Spark or Similar Service
+
+**Proposal**
+
+Instead of aggregating in ClickHouse we should aggregate our replays in a stateful service such as Apache Spark before writing the final result to ClickHouse.
+
+TODO: Josh to fill in specifics.
+
+**Drawbacks**
+
+- Replays are not available until they have finished.
+- There are no existing Apache Spark installations within the Sentry org.
+- The Replays team is not large enough or experienced enough to manage a Spark installation.
+  - This would require another team assuming the burden for us.
+  - Otherwise, additional budget would need to be allocated to the Replays team to hire outside experts.
+
+**Questions**
+
+### 7. Upgrade ClickHouse Version to Utilize Experimental Features
+
+**Proposal**
+
+Upgrading to a newer version of ClickHouse will enable us to use experimental features such as "Live View" and "Window View".
+
+**Drawbacks**
+
+- Would require Sentry to manage the ClickHouse installation.
+  - Feature is not present in a stable release.
+  - Feature is not present on any version of ClickHouse in Altinity.
+
+**Questions**
+
+### 8. Use an Alternative OLAP Database
+
+**Proposal**
+
+OLAP Databases such as Apache Pinot support upserts which appear to be a key requirement for the Replays product. An aggregation state schema can be defined for merging columns in real-time https://docs.pinot.apache.org/basics/data-import/upsert.
+
+**Drawbacks**
+
+- Changes to the data model would be necessary. We will not be able to aggregate array columns.
+- Pinot has ordering constraints.
+  - If segments arrive after their successor has already been ingested the aggregated state will not contain those rows.
+- There are no existing Apache Pinot installations within the Sentry org.
+- The Replays team is not large enough or experienced enough to manage a Pinot installation.
+  - This would require another team assuming the burden for us.
+  - Otherwise, additional budget would need to be allocated to the Replays team to hire outside experts.
+- We need to re-write our application logic for querying the datastore.
+- Migration pains.
+
+**Questions**
+
+### 9. Use an Alternative OLTP Database
+
+**Proposal**
+
+OLTP databases such as PostgreSQL and AlloyDB support updates which appear to be a key requirement for the Replays product.
+
+**Drawbacks**
+
+- Race conditions will require single-threaded processing of replay events.
+- Duplicate messages will necessitate ordering requirements.
+- Always a possibility for dropped and duplicated data regardless of safe guards.
+- AlloyDB is still in developer preview on Google Cloud.
+- We need to re-write our application logic for querying the datastore.
+- Migration pains.
+
+**Questions**
+
```

Comment:
I think one other idea would be to do our own managing of materialized views. how this would work:

1. we have a cron that runs every 24 hours, and selects all finished replays that occured in the last 24 hours, aggregates them, and writes them to a materialized view table. when we query replays, we query both the "last 24 hours table" and "beyond 1 day pre-aggregated table" and join the results.

This could be extended to run every hour or quicker, to try and pre-aggregate all completed replays. someone with more clickhouse knowledge could perhaps expand upon this.

Reply:
Added.